### DISCUSSION TITLE
Training bottomup-id model is hard.

### DISCUSSION THREAD

**Discussion by olivier-cuttlefish** [talmolab/sleap#1567] [Help!]
I am having difficulties training a bottomup-id model (difficulties as in it doesn't converge) whereas an identitical bottomup model (except for the heads parameters) can easily be trained, with the same annotations (~100 annotations)
Here is an example of the evolution of losses while training with bottomup-id. 
![image](https://github.com/talmolab/sleap/assets/114718516/d1371cdb-c0a8-4cc8-a504-b60147d84ca1)

I am suspecting that there is an issue with the relative loss between the body parts confidence maps and the class maps. This training was done with lossweight 2:1 in favor of the body parts confidence maps. As a result of the training (I cannot publish example pictures of the data on github, but am willing to share it in private if necessary), the bodyparts confidence maps are really bad and do not manage to predict points. As for the class maps, it seems that they are drawn correctly around the animals (there are two), but am having difficulties interpreting it. I know that this implementation is coming from idtracker.ai (correct me if I'm wrong), but I don't know how it is implemented in SLEAP, and I could not find information about it on the github nor documentation. I have just started a new training with lossweight 10:1 in favor to confidence maps, but I am not sure I am tackling properly the issue.
Would you have any recommendations on how to improve the training ? And details about how this id tracking is implemented in SLEAP?

This is my latest model configuration: 
```
"model": {
"backbone": {
"leap": null,
"unet": {
"stem_stride": null,
"max_stride": 64,
"output_stride": 2,
"filters": 16,
"filters_rate": 2.0,
"middle_block": true,
"up_interpolate": true,
"stacks": 1
},
"hourglass": null,
"resnet": null,
"pretrained_encoder": null
},
"heads": {
"single_instance": null,
"centroid": null,
"centered_instance": null,
"multi_instance": null,
"multi_class_bottomup": {
"confmaps": {
"part_names": null,
"sigma": 5.0,
"output_stride": 2,
"loss_weight": 2.0,
"offset_refinement": false
},
"class_maps": {
"classes": [
"1",
"2"
],
"sigma": 15.0,
"output_stride": 4,
"loss_weight": 1.0
}
},
"multi_class_topdown": null
},
"base_checkpoint": null
},
"optimization": {
"preload_data": true,
"augmentation_config": {
"rotate": true,
"rotation_min_angle": -180.0,
"rotation_max_angle": 180.0,
"translate": false,
"translate_min": -5,
"translate_max": 5,
"scale": true,
"scale_min": 0.9,
"scale_max": 1.1,
"uniform_noise": true,
"uniform_noise_min_val": 0.0,
"uniform_noise_max_val": 10.0,
"gaussian_noise": true,
"gaussian_noise_mean": 5.0,
"gaussian_noise_stddev": 1.0,
"contrast": true,
"contrast_min_gamma": 0.5,
"contrast_max_gamma": 2.0,
"brightness": true,
"brightness_min_val": 0.0,
"brightness_max_val": 10.0,
"random_crop": false,
"random_crop_height": 256,
"random_crop_width": 256,
"random_flip": true,
"flip_horizontal": false
},
"online_shuffling": true,
"shuffle_buffer_size": 128,
"prefetch": true,
"batch_size": 4,
"batches_per_epoch": null,
"min_batches_per_epoch": 200,
"val_batches_per_epoch": null,
"min_val_batches_per_epoch": 10,
"epochs": 100,
"optimizer": "adam",
"initial_learning_rate": 0.0001,
"learning_rate_schedule": {
"reduce_on_plateau": true,
"reduction_factor": 0.5,
"plateau_min_delta": 1e-06,
"plateau_patience": 10,
"plateau_cooldown": 3,
"min_learning_rate": 1e-08
},
"hard_keypoint_mining": {
"online_mining": false,
"hard_to_easy_ratio": 2.0,
"min_hard_keypoints": 2,
"max_hard_keypoints": null,
"loss_scale": 5.0
},
"early_stopping": {
"stop_training_on_plateau": true,
"plateau_min_delta": 1e-06,
"plateau_patience": 20
}
},
```

Thank you very much for your support.

**Comment by olivier-cuttlefish** [comment]
For people that might have ran into similar issues, I fixed the training by using a loss weight of for confidence maps, and 0.001 for class maps. As can be seen in my training log, the loss of ClassMaps are disproportionately high and needs to be scaled to the ConfMaps loss.

However, when trying to then run inference using that model, the model generate only one instance (out of two), picking only the best predicted body parts, which results in a single instance which is often a mix of the two animals in my pictures. I guess this is unintended behavior but don't know yet how to fix it.

**Comment by Marielamend** [comment]
Hello @olivier-cuttlefish!
So sorry looks like your inquiry fell through the cracks. Do you still have this issue? If you still have this issue let us know and/or if you have any further questions.

Have a great day,
Mariela

**Reply by olivier-cuttlefish** [reply]
Hi @Marielamend , thank you for your answer! I have not fixed this issue yet, and just went for a workaround (bottomup then tracking). However, my lab and I would be very interested in having those model-id models working, as we often make use of mutlti-individuals tracking. My issue is still at the same stage of my last message in this thread. Do you know how to properly configure the bottomup-id model? And also why I had this weird behavior described in my last message? 

Have a nice week-end,
Olivier

**Reply by Marielamend** [reply]
Hello @olivier-cuttlefish!,

Sorry, you have been having some issues with this. 

The best you can do is tune the loss weights for confidence maps vs class maps just as you’ve been doing. 

If you can provide us with the training_log.csv from your model folder, this can allow us to visualize the loss curves from training to see where the 10:1 variant is at in terms of balance. 

To tweak configurations for bottom-up: taken from similar issue #1387:

> - Try decreasing the input scale to 0.5. This seems counterintuitive since you're losing resolution, but the images you have are fairly large and the features fairly coarse, so it might be a better starting point without significant loss of performance.
> - Decrease the filter rate to 1.5 and increase the filters to 48 to balance the representational capacity in favor of larger scales
 

Lastly, you can also try the top-down approach which has two stages where the first stage consists of the neural network detecting the centroid (the anchor point) of the animal first!

Have a great day,
Mariela
